### ISSUE TITLE
:bind defeats :after

### ISSUE THREAD

**Issue by tarsius** [jwiegley/use-package#705]
I had this in my init file:

```lisp
(use-package bicycle :after outline)
```

This expands to something involving `(require 'bicycle)`, but only after `outline` is loaded.

But then I added my personal key bindings:

```lisp
(use-package bicycle
  :after outline
  :bind (:map outline-minor-mode-map
              ([C-tab] . bicycle-cycle)
              ([S-tab] . bicycle-cycle-global)))
```

And now the expansion involves `autoload` and the `require` is gone. It seems to me that `:after` should be enough to cause the package to be loaded `:after` the other package, regardless of what other keywords are involved.

I was able to force the `require` to reappear by adding `:demand`, which resulted in exactly the behavior I would have expected *without* that addition: `bicycle`'s loading was once more deferred until after `outline` was loaded. However, that does not match the documented behavior of `:demand`, which claims that it "prevent[s] deferred loading in all cases". In this case it does not prevent deferred loading, it enables it.

**Comment by tarsius** [comment]
It seems this could be fixed by having `:after` affect the expansion of some other keywords the same as `:defered` does, i.e. it should prevent the code-paths involving `autoload` from being used.

**Comment by jwiegley** [comment]
@tarsius Your assessment sounds about right to me.

**Comment by unhammer** [comment]
Not sure if this is related, but `:if` seems to get evaluated too eagerly too, so e.g.
```
(use-package orgit
  :ensure t
  :after org
  :if (string> (org-version) "8.3.3")
  )
```
where I suppose `(org-version)` autoloads `org`. It'd be nice if `:after`'s deferral included what's in `if`.

**Comment by moyamo** [comment]
I think both behaviours are useful. Maybe the `:after` keyword should be split

- `:demand-after x` Load the package immediately after `x` is loaded
- `:defer-after x` Prevent the loading of the package until after `x` is loaded.

Of course the former is equivalent to `:demand :after x` and the latter equivalent to `(:defer|:bind|etc.) :after x`, so maybe it's just the documentation that needs changing.

**Comment by progfolio** [comment]
@moyamo: I'm curious if you could a provide an example of how one would achieve the :defer-after behavior with the current version of use-package. I've just run into this bug when moving my key bindings to their respective use-package declarations.

I'm running into trouble because elfeed's use package form is before general's. However, I have elfeed specified to load :after (general), and am using the :general keyword provided by [general.el](https://github.com/noctuid/general.el), yet I'm getting this use-package error:

`Error (use-package): Failed to parse package elfeed: use-package: Unrecognized keyword: :general`

Here's a minimal version of my config that illustrates the problem:

    ;; -*- lexical-binding: t; -*-

    (setq package-archives
          '(("gnu"   . "https://elpa.gnu.org/packages/")
            ("melpa" . "https://melpa.org/packages/")
            ("org"   . "https://orgmode.org/elpa/")))

    (require 'package)
    (package-initialize)

    (unless (package-installed-p 'use-package)
      (package-refresh-contents)
      (package-install 'use-package))

    (require 'use-package)
    (use-package use-package-core
      :ensure nil
      :init (setq use-package-verbose t
                  use-package-expand-minimally nil
                  use-package-always-defer t
                  use-package-always-ensure t))

    (use-package elfeed
      :after (general)
      :general
      ;;this is a macro that is defined in the full init.
      ;;use-package never gets this far, though...
      (general-global-application
        "e"    'elfeed)
      )

    (use-package general
      :demand t
      :config
      (general-override-mode)
      (general-auto-unbind-keys))

**Comment by QiangF** [comment]
why not just move use-package general before use-package elfeed?

**Comment by waymondo** [comment]
@progfolio Your issue is when `general` is loaded via `use-package`, it adds support for the use-package keyword of `:general`. Since `general` has not been loaded by the time the `use-package` form for `elfeed` is evaluated, `use-package` doesn't know what to do with the `:general` keyword. You'll need to load `general` before you can use `:general`.

**Comment by progfolio** [comment]
Ah! A fundamental misunderstanding on my part. My apologies for the noise on this thread. Thank you for the explanation, @waymondo.

**Comment by tarsius** [comment]
> Your assessment sounds about right to me.

@jwiegley I had another and I am afraid I did not gain as deep an understanding as the last time I did (assuming that wasn't an illusion back then in the first place). Could you please fix this? I feel it would take me a very long time to learn enough to *maybe* eventually be able to fix it. You would certainly have an easier time doing it.

**Comment by jwiegley** [comment]
Sure, @tarsius.

**Comment by jwiegley** [comment]
@tarsius Ok, I've looked into this further, and now think that the current behavior is what's intended: The purpose of `:after` is to defer the `use-package` declaration's effect until after the mentioned package has been loaded. In your case, it defers binding of the keys until after `outline` has loaded. The _deferred nature_ of those bindings, however, is still in effect, because you didn't use `:demand t`. Is that what you had assumed would happen? Here's the behavior in that case:

```lisp
(eval-after-load 'outline
  '(progn
     (require 'bicycle nil nil)
     (bind-keys :package bicycle :map outline-minor-mode-map
                ([C-tab]
                 . bicycle-cycle)
                ([S-tab]
                 . bicycle-cycle-global))))
```

So, presently `:after` means: Do whatever `use-package` would have done, but not until after the given package is loaded. It doesn't alter the meaning of `:demand` or `:defer` in any way. I feel that if we did start doing that, things would get more confusing, beacuse then just introducing `:after` would change the meaning.

**Comment by tarsius** [comment]
I am now coming to the same conclusion but with a caveat.

Lets first look at the two expansions again. This is what I originally found surprising:

```emacs-lisp
(use-package bicycle
  :after outline
  :bind ...)

(eval-after-load 'outline
  '(progn
     (unless (fboundp 'bicycle-cycle)
       (autoload #'bicycle-cycle "bicycle" nil t))
     (bind-keys ...)))
```

This is what we get when adding `:demand t` and what I was expecting for `:demand nil`:

```emacs-lisp
(use-package bicycle
  :after outline
  :demand t
  :bind ...)

(eval-after-load 'outline
  '(progn
     (require 'bicycle nil t)
     (bind-keys ...)))
```

I believe the reason that I looked into this at all was that I believed that the `(use-package bicylce ...)` caused `outline` to be loaded and that adding `:demand t` caused that to no longer be the case. Both of these observations were wrong. (The reason that `outline` did get loaded (and still does (and I don't really mind that (because I only minded not understanding what was going on))) is that I have added `outline-minor-mode` to `emacs-lisp-mode`, which always gets called at the end of the init process because `*scrach*`'s mode derives from that.)

Of course evaluating a `autoload` does not cause the library with the real definition to be loaded.

----

And now for the caveat:

```
:after           Defer loading of a package until after any of the named
                 features are loaded.
:demand          Prevent deferred loading in all cases.
```

I still object to the use of "in all cases" in this case. :stuck_out_tongue_winking_eye: I think this documentation is at least misleading. An addition in this spirit could help:

```
:after           Defer loading of a package until after any of the named
                 features are loaded, without actually causing the package to be loaded
                 after that has happened. Whether it is loaded then is still subject to
                 other keywords, such as `:demand` and `:defer`.
:demand          Prevent deferred loading. Note that this does not override the effect
                 of `:after`.
```

The wording needs some work, but I hope I at least got across what additional information I think should be provided here.

**Comment by tarsius** [comment]
#803 is related.

**Comment by jwiegley** [comment]
Yes, I think you're right that the documentation should be updated. I find the original descriptions confusing too. :)

**Comment by jwiegley** [comment]
@tarsius See if that PR isn't more to your liking...
